File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();
        if (n < 3) return {};
        
        vector<vector<int>> ans;
        sort(nums.begin(), nums.end());

        for (int i = 0; i <= n - 3; i++) {
            // the fixed value is the same as the previous one, we ignore it for deduplication
            if (i != 0 && nums[i] == nums[i - 1]) continue;

            // solve the two sum problem with two pointers
            int left = i + 1;
            int right = n - 1;
            int target = -nums[i];

            while (left < right) {
                int s = nums[left] + nums[right];
                if (s > target) {
                    right--;
                } else if (s < target) {
                    left++;
                } else {
                    ans.push_back({ nums[i], nums[left], nums[right] });
                    left++;
                    right--;
                    // avoid the duplicates
                    while (left < right && nums[left - 1] == nums[left]) left++;
                    while (left < right && nums[right] == nums[right + 1]) right--;
                }
            }
        }

        return ans;
    }
};

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        if len(nums) < 3:
            return []
        
        nums.sort()

        ans = []
        for i in range(len(nums)):
            if i > 0 and nums[i - 1] == nums[i]:
                continue
            # two pointers
            target = -nums[i]
            j = i + 1
            k = len(nums) - 1
            while j < k:
                added = nums[j] + nums[k]
                if added < target:
                    j += 1
                elif added > target:
                    k -= 1
                else:
                    ans.append([nums[i], nums[j], nums[k]])
                    j += 1
                    k -= 1
                    while j < k and nums[j - 1] == nums[j]:
                        j += 1
                    while j < k and nums[k] == nums[k + 1]:
                        k -= 1
        return ans